addons/purchase/migrations/9.0.1.2/post-migration.py

```diff
@@ -92,9 +96,188 @@ def account_properties(cr):
             """)
 
 
+@openupgrade.logging()
+def set_po_line_amounts(env):
+    """We replicate here the code of the compute function and set the values
+    finally via SQL for avoiding the trigger of the rest of the computed
+    fields that depends on these fields.
+
+    Replaces code from https://bit.ly/2rgpTTV
+    """
+    lines = env['purchase.order.line'].search([])
+    for line in lines:
+        taxes = line.taxes_id.compute_all(
+            line.price_unit, line.order_id.currency_id, line.product_qty,
+            product=line.product_id, partner=line.order_id.partner_id)
+        env.cr.execute("""
+            UPDATE purchase_order_line
+            SET price_tax = %s,
+                price_total = %s,
+                price_subtotal = %s
+            WHERE id = %s""", (
+                taxes['total_included'] - taxes['total_excluded'],
+                taxes['total_included'],
+                taxes['total_excluded'],
+                line.id,
+            )
+        )
+
+
+@openupgrade.logging()
+def set_po_line_computed_rest(env):
+    """Emulate the computation of the rest of the computed fields through
+    equivalent SQL queries.
+    """
+    uom_precision = env['decimal.precision'].precision_get('Product Unit of Measure')
+    # Fail if there are UoM mismatches
+    # Replace https://bit.ly/2s1ePtG
+    openupgrade.logged_query(
+        env.cr,
+        """
+            SELECT
+                pol.id AS order_line_id,
+                pol_u.id AS order_line_uom_id,
+                pol_u.category_id AS order_line_uom_category_id,
+                pol.product_id AS order_line_product_id,
+                ail.id AS invoice_line_id,
+                ail_u.id AS invoice_line_uom_id,
+                ail_u.category_id AS invoice_line_uom_category_id,
+                ail_pp.id AS invoice_line_product_id,
+                ail_pt.name AS invoice_line_product_template_name
+            FROM purchase_order_line pol
+                JOIN account_invoice_line ail ON
+                    pol.id = ail.purchase_line_id
+                JOIN product_product ail_pp ON ail_pp.id = ail.product_id
+                JOIN product_template ail_pt ON ail_pt.id = ail_pp.product_tmpl_id
+                JOIN account_invoice ai ON ai.id = ail.invoice_id
+                JOIN product_uom ail_u on (ail_u.id = ail.uom_id)
+                JOIN product_uom pol_u on (pol_u.id = pol.product_uom)
+            WHERE
+                ail_u.category_id != pol_u.category_id
+        """
+    )
+    uom_category_mismatches = env.cr.dictfetchall()
+    if uom_category_mismatches:
+        raise Exception(
+            "Found these mismatching UoM in related purchase.order.line "
+            "and account.invoice.line records: %s" %
+            pformat(uom_category_mismatches))
+    # Replace https://bit.ly/363UnqR
```

Comment:
I really find the original links, even if they are too long much more readable. Maybe mark them with noqa or split them up on a slash forward?

Reply:
Replaced by full URLs